kernel-open/common/inc/nv-nanos.h

```diff
@@ -507,8 +519,13 @@ typedef struct nvidia_event
 #define vmalloc(size)       kmalloc(size, 0)
 #define vzalloc(size)       kzalloc(size, 0)
 #define ksize(p)            objcache_from_object(u64_from_pointer(p), PAGESIZE_2M)->pagesize
-#define is_vmalloc_addr(p)  false
-#define vfree               kfree
+#define is_vmalloc_addr(p)  (objcache_from_object(u64_from_pointer(p), PAGESIZE_2M) == INVALID_ADDRESS)
+#define vfree(p) do {                               \
+    uvm_vmalloc_hdr_t *hdr;                         \
+    hdr = container_of(p, uvm_vmalloc_hdr_t, ptr);  \
+    NV_KFREE(p, hdr->alloc_size);                   \
+} while (0)
+
```

Comment:
This won't work if memory is allocated from outside the nvidia-uvm module, because it won't have the uvm_vmalloc_hdr_t header.
To deal with the vfree() call in linux_nvswitch.c, I would not use vmalloc/vfree at all in the nvswitch code; instead, I would use use another Nanos kernel heap, i.e. `get_kernel_heaps()->malloc`, which can take -1 as size argument when freeing memory. So, _nvswitch_os_malloc() will be as below:
```
    void *ptr = allocate(get_kernel_heaps()->malloc, size);

    if (ptr == INVALID_ADDRESS)
        return NULL;

    return ptr;
```
and _nvswitch_os_free() will be as below:
```
    if (!ptr)
        return;

    deallocate((get_kernel_heaps()->malloc), ptr, -1ull);
```

The vfree() macro will only be called from the nvidia-uvm module, where we can easily extract the size parameter from the uvm_vmalloc_hdr_t header.